Let JIT unpickler to accept CUDA DataPtr from read_record_

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46791 Add cuda_ipc channel to TensorPipe
* **#46827 Let JIT unpickler to accept CUDA DataPtr from read_record_**

TensorPipe RPC agent uses JIT pickler/unpickler to serialize/deserialize
tensors. Instead of saving tensors to a file, the agent can directly
invoke `cudaMemcpy` to copy tensors from the sender to the receiver
before calling into JIT unpickler. As a result, before unpickling,
the agent might already have allocated tensors and need to pass
them to the JIT unpickler. Currently, this is done by providing a
`read_record` lambda to unpickler for CPU tensors, but this is
no longer sufficient for zero-copy CUDA tensors, as the unpickler
always allocate the tensor on CPU.

To address the above problem, this commit introduces a `use_storage_device`
flag to unpickler ctor. When this flag is set, the unpickler will
use the device from the `DataPtr` returned by the `read_record`
lambda to override the pickled device information and therefore
achieves zero-copy.

Differential Revision: [D24533218](https://our.internmc.facebook.com/intern/diff/D24533218)

mrshenli committed:
Let JIT unpickler to accept CUDA DataPtr from read_record_

[ghstack-poisoned]

mrshenli committed:
Update on "Let JIT unpickler to accept CUDA DataPtr from read_record_"


TensorPipe RPC agent uses JIT pickler/unpickler to serialize/deserialize
tensors. Instead of saving tensors to a file, the agent can directly
invoke `cudaMemcpy` to copy tensors from the sender to the receiver
before calling into JIT unpickler. As a result, before unpickling,
the agent might already have allocated tensors and need to pass
them to the JIT unpickler. Currently, this is done by providing a
`read_record` lambda to unpickler for CPU tensors, but this is
no longer sufficient for zero-copy CUDA tensors, as the unpickler
always allocate the tensor on CPU.

To address the above problem, this commit introduces a `use_storage_device`
flag to unpickler ctor. When this flag is set, the unpickler will
use the device from the `DataPtr` returned by the `read_record`
lambda to override the pickled device information and therefore
achieves zero-copy.

[ghstack-poisoned]

mrshenli commented:
Hey @wanchaol, this is roughly what I need. Let me know if this looks OK to you. Thanks!

lw commented on torch/csrc/jit/serialization/unpickler.cpp:
```diff
@@ -417,6 +417,12 @@ PickleOpCode Unpickler::readInstruction() {
+
+      if (use_storage_device_) {
+        options = options.device(storage.device());
+        device = storage.device();
```
This means that we will still do the `tensor = tensor.to(device)` call later down, though, right? Are we sure this is a no-op if `tensor` is already on `device`?

mrshenli replied:
Yep, it should be a no-op IIUC. I can make it more explicit by avoid that `.to()` to make it safer.

wanchaol reviewed (commented):
Logic looks good to me. But could you explain a bit more about the motivation on why we need to pass it to the unpickler if we are performing zero-copy? is it just wrapping the StoragePtr back to a tensor?

Also, afaik, for cuda tensors, the pickler will pickle the location to "cuda" and when we create the tensor back in the unpickler, if the storagePtr is from a cuda memory (i.e. read by tensorpipe's `read_record`), wouldn't that already be a cuda tensor?

mrshenli commented:
Hey @wanchaol, thanks for the comment.

> But could you explain a bit more about the motivation on why we need to pass it to the unpickler if we are performing zero-copy? 

I initially thought about adding this to the pickler, but I later realized that this is not possible, as the pickler might not know where the pickled tensor will be unpickled to. It could be on a different device.

> is it just wrapping the StoragePtr back to a tensor?

Yep, just let the `StoragePtr` dictate the device of the tensor. IIUC, the `device` we currently got in the unpickler is determined at the pickling time, which might be different from the device returned by the `read_record_` function.

> Also, afaik, for cuda tensors, the pickler will pickle the location to "cuda" and when we create the tensor back in the unpickler, if the storagePtr is from a cuda memory (i.e. read by tensorpipe's read_record), wouldn't that already be a cuda tensor?

Yes, it will be a CUDA tensor but on a different GPU. One example would be:

```python
options.set_device_map("worker1", {0: 1})
init_rpc("worker0", options)
x = torch.zeros(2).to(0)
rpc_sync("worker1", torch.add, args=(x, 1))
```

In the above example, TensorPipe will copy `x` from `worker0` `cuda:0` to `worker1` `cuda:1` using `cudaMemcpy`. So, when we unpickle the tensor on `worker1` device will be `cuda:0`, but the storage will be `cuda:1`.

wanchaol reviewed (approved):
looks good to me.

mrshenli committed:
Update on "Let JIT unpickler to accept CUDA DataPtr from read_record_"


TensorPipe RPC agent uses JIT pickler/unpickler to serialize/deserialize
tensors. Instead of saving tensors to a file, the agent can directly
invoke `cudaMemcpy` to copy tensors from the sender to the receiver
before calling into JIT unpickler. As a result, before unpickling,
the agent might already have allocated tensors and need to pass
them to the JIT unpickler. Currently, this is done by providing a
`read_record` lambda to unpickler for CPU tensors, but this is
no longer sufficient for zero-copy CUDA tensors, as the unpickler
always allocate the tensor on CPU.

To address the above problem, this commit introduces a `use_storage_device`
flag to unpickler ctor. When this flag is set, the unpickler will
use the device from the `DataPtr` returned by the `read_record`
lambda to override the pickled device information and therefore
achieves zero-copy.

Differential Revision: [D24533218](https://our.internmc.facebook.com/intern/diff/D24533218)

[ghstack-poisoned]

mrshenli committed:
Update on "Let JIT unpickler to accept CUDA DataPtr from read_record_"


TensorPipe RPC agent uses JIT pickler/unpickler to serialize/deserialize
tensors. Instead of saving tensors to a file, the agent can directly
invoke `cudaMemcpy` to copy tensors from the sender to the receiver
before calling into JIT unpickler. As a result, before unpickling,
the agent might already have allocated tensors and need to pass
them to the JIT unpickler. Currently, this is done by providing a
`read_record` lambda to unpickler for CPU tensors, but this is
no longer sufficient for zero-copy CUDA tensors, as the unpickler
always allocate the tensor on CPU.

To address the above problem, this commit introduces a `use_storage_device`
flag to unpickler ctor. When this flag is set, the unpickler will
use the device from the `DataPtr` returned by the `read_record`
lambda to override the pickled device information and therefore
achieves zero-copy.

Differential Revision: [D24533218](https://our.internmc.facebook.com/intern/diff/D24533218)

[ghstack-poisoned]

mrshenli committed:
Update on "Let JIT unpickler to accept CUDA DataPtr from read_record_"


TensorPipe RPC agent uses JIT pickler/unpickler to serialize/deserialize
tensors. Instead of saving tensors to a file, the agent can directly
invoke `cudaMemcpy` to copy tensors from the sender to the receiver
before calling into JIT unpickler. As a result, before unpickling,
the agent might already have allocated tensors and need to pass
them to the JIT unpickler. Currently, this is done by providing a
`read_record` lambda to unpickler for CPU tensors, but this is
no longer sufficient for zero-copy CUDA tensors, as the unpickler
always allocate the tensor on CPU.

To address the above problem, this commit introduces a `use_storage_device`
flag to unpickler ctor. When this flag is set, the unpickler will
use the device from the `DataPtr` returned by the `read_record`
lambda to override the pickled device information and therefore
achieves zero-copy.

Differential Revision: [D24533218](https://our.internmc.facebook.com/intern/diff/D24533218)

[ghstack-poisoned]
